# Ultralytics Framework for YOLOv10 Training

In [1]:
import torch

# Number of GPUs available
num_gpus = torch.cuda.device_count()

print(f"Number of available GPUs: {num_gpus}")


Number of available GPUs: 4


# Yolov10m Model Training on:
- Carla_Labelling-v5 dataset
- Train: 70% = 4485 Images
- Validation: 20% = 1274 Images
- Test: 10% = 641 Images

- Hyperparameters are on automatic settings using Ultralytics package

In [2]:
from ultralytics import YOLO

# Load YOLOv10n model from scratch
model = YOLO("yolov10m.pt")

100%|██████████| 32.1M/32.1M [00:00<00:00, 77.8MB/s]


# Validate the pretrained weights by running inference on an image.

In [3]:
# Run batched inference on a list of images
results = model(["/home/sur06423/project/1.jpg"])  # return a list of Results objects

# Process results list
for result in results:
    boxes = result.boxes  # Boxes object for bounding box outputs
    masks = result.masks  # Masks object for segmentation masks outputs
    keypoints = result.keypoints  # Keypoints object for pose outputs
    probs = result.probs  # Probs object for classification outputs
    obb = result.obb  # Oriented boxes object for OBB outputs
    result.show()  # display to screen
    result.save(filename="result.jpg")  # save to disk


0: 448x640 4 horses, 67.3ms
Speed: 6.1ms preprocess, 67.3ms inference, 25.8ms postprocess per image at shape (1, 3, 448, 640)


Error: no "view" rule for type "image/png" passed its test case
       (for more information, add "--debug=1" on the command line)


# Train the model using the following guidelines:
- Link for various arguments: https://docs.ultralytics.com/modes/train/#train-settings 

In [4]:
import os

# Set library paths
os.environ['LD_LIBRARY_PATH'] = '/usr/lib/xorg-nvidia-525.116.04/lib/x86_64-linux-gnu:' + os.environ.get('LD_LIBRARY_PATH', '')
os.environ['LD_LIBRARY_PATH'] = '/usr/lib/xorg/lib/x86_64-linux-gnu:' + os.environ['LD_LIBRARY_PATH']
os.environ['LD_LIBRARY_PATH'] = '/usr/lib/xorg-nvidia-535.113.01/lib/x86_64-linux-gnu:' + os.environ['LD_LIBRARY_PATH']

# Verify the update
print(os.environ['LD_LIBRARY_PATH'])


/usr/lib/xorg-nvidia-535.113.01/lib/x86_64-linux-gnu:/usr/lib/xorg/lib/x86_64-linux-gnu:/usr/lib/xorg-nvidia-525.116.04/lib/x86_64-linux-gnu:/home/sur06423/miniconda3/envs/yolov_env/lib/python3.9/site-packages/cv2/../../lib64:


In [5]:
# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
model.train(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", # 	Path to the dataset configuration file (e.g., coco8.yaml). This file contains dataset-specific parameters, including paths to training and validation data, class names, and number of classes.
            epochs=100, # Total number of training epochs. Each epoch represents a full pass over the entire dataset. Adjusting this value can affect training duration and model performance.
            patience=10, # Number of epochs to wait without improvement in validation metrics before early stopping the training. Helps prevent overfitting by stopping training when performance plateaus.
            batch=128, # Batch size, with three modes: set as an integer (e.g., batch=16), auto mode for 60% GPU memory utilization (batch=-1), or auto mode with specified utilization fraction (batch=0.70).
            imgsz=640, # Target image size for training. All images are resized to this dimension before being fed into the model. Affects model accuracy and computational complexity.
            save=True, # Enables saving of training checkpoints and final model weights. Useful for resuming training or model deployment.
            save_period = 1, # Frequency of saving model checkpoints, specified in epochs. A value of -1 disables this feature. Useful for saving interim models during long training sessions.
            device=[0,1,2,3], # Specifies the computational device(s) for training: a single GPU (device=0), multiple GPUs (device=0,1), CPU (device=cpu), or MPS for Apple silicon (device=mps).
            workers=6, # Number of worker threads for data loading (per RANK if Multi-GPU training). Influences the speed of data preprocessing and feeding into the model, especially useful in multi-GPU setups.
            project="/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m", # Name of the project directory where training outputs are saved. Allows for organized storage of different experiments.
            pretrained = True, # Determines whether to start training from a pretrained model. Can be a boolean value or a string path to a specific model from which to load weights. Enhances training efficiency and model performance.
            optimizer = "auto", # Choice of optimizer for training. Options include SGD, Adam, AdamW, NAdam, RAdam, RMSProp etc., or auto for automatic selection based on model configuration. Affects convergence speed and stability.
            verbose = True, # Enables verbose output during training, providing detailed logs and progress updates. Useful for debugging and closely monitoring the training process.
            seed = 0, # Sets the random seed for training, ensuring reproducibility of results across runs with the same configurations.
            amp = True, # Enables Automatic Mixed Precision (AMP) training, reducing memory usage and possibly speeding up training with minimal impact on accuracy.
            profile =True, # Enables profiling of ONNX and TensorRT speeds during training, useful for optimizing model deployment.
            val=True, # Enables validation during training, allowing for periodic evaluation of model performance on a separate dataset.
            plots = True, #Generates and saves plots of training and validation metrics, as well as prediction examples, providing visual insights into model performance and learning progression.                
)

New https://pypi.org/project/ultralytics/8.3.13 available 😃 Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:1 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:2 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:3 (NVIDIA RTX A6000, 48677MiB)
engine/trainer: task=detect, mode=train, model=yolov10m.pt, data=/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml, epochs=100, time=None, patience=10, batch=128, imgsz=640, save=True, save_period=1, cache=False, device=[0, 1, 2, 3], workers=6, project=/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0

Overriding model.yaml nc=80 with nc=4

                   from  n    params  module                                       arguments                     
  0                  -1  1      1392  ultralytics.nn.modules.conv.Conv             [3, 48, 3, 2]                 
  1                  -1  1     41664  ultralytics.nn.modules.conv.Conv             [48, 96, 3, 2]                
  2                  -1  2    111360  ultralytics.nn.modules.block.C2f             [96, 96, 2, True]             
  3                  -1  1    166272  ultralytics.nn.modules.conv.Conv             [96, 192, 3, 2]               
  4                  -1  4    813312  ultralytics.nn.modules.block.C2f             [192, 192, 4, True]           
  5                  -1  1     78720  ultralytics.nn.modules.block.SCDown          [192, 384, 3, 2]              
  6                  -1  4   3248640  ultralytics.nn.modules.block.C2f             [384, 384, 4, True]           
  7                  -1  1    228672  ultralytics

100%|██████████| 6.25M/6.25M [00:00<00:00, 116MB/s]


AMP: checks passed ✅


train: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/train/labels.cache... 4485 images, 430 backgrounds, 0 corrupt: 100%|██████████| 4485/4485 [00:00<?, ?it/s]
val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/valid/labels.cache... 1274 images, 124 backgrounds, 0 corrupt: 100%|██████████| 1274/1274 [00:00<?, ?it/s]


Plotting labels to /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 129 weight(decay=0.0), 142 weight(decay=0.001), 141 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 24 dataloader workers
Logging results to /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      1/100      17.3G      2.499      7.669      2.088          5        640: 100%|██████████| 36/36 [00:18<00:00,  1.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.71it/s]


                   all       1274       7377      0.514      0.377      0.349        0.2

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      2/100      17.4G      2.421      2.285      2.015         17        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.81it/s]


                   all       1274       7377      0.273      0.231      0.127     0.0629

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      3/100      17.4G      2.537      2.103       2.07         37        640: 100%|██████████| 36/36 [00:16<00:00,  2.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.90it/s]


                   all       1274       7377     0.0557    0.00866    0.00558    0.00211

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      4/100      17.3G      2.544      2.008      2.102          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.439      0.327      0.256      0.142

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/100      17.3G      2.474      1.855      2.033          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.595       0.44      0.426      0.243

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      6/100      17.3G        2.4      1.689      1.983         26        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.645      0.448      0.477      0.278

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/100      17.5G      2.338      1.639      1.978          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.671      0.445      0.471       0.27

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/100      17.4G      2.315      1.548       1.93         24        640: 100%|██████████| 36/36 [00:16<00:00,  2.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.89it/s]


                   all       1274       7377       0.72      0.508      0.558      0.324

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/100      17.4G       2.21      1.484      1.944          1        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.86it/s]


                   all       1274       7377      0.729      0.491      0.538      0.324

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/100      17.3G       2.23      1.397      1.908          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377      0.742      0.503      0.562      0.336

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     11/100      17.3G      2.267      1.403        1.9         14        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.80it/s]


                   all       1274       7377       0.76      0.529      0.586      0.364

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     12/100      17.4G      2.206      1.624      1.867          1        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.781      0.539      0.597      0.362

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     13/100      17.5G      2.195      1.364      1.906         11        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.98it/s]


                   all       1274       7377       0.78      0.536      0.602      0.373

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     14/100      17.4G      2.163      1.373      1.881         14        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.03it/s]


                   all       1274       7377      0.797      0.547      0.614      0.383

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     15/100      17.4G      2.138      1.299      1.861         24        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.98it/s]


                   all       1274       7377      0.772      0.557      0.617      0.382

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/100      17.4G      2.106      1.273      1.859          4        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.03it/s]


                   all       1274       7377      0.796      0.565      0.626      0.387

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     17/100      17.4G      2.086      1.259      1.877         15        640: 100%|██████████| 36/36 [00:16<00:00,  2.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.71it/s]


                   all       1274       7377      0.824      0.554      0.625      0.392

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     18/100      17.3G      2.055       1.26      1.874          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377       0.81      0.554      0.627      0.389

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     19/100      17.4G      2.112      1.246      1.854         12        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.816      0.551      0.627      0.396

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/100      17.4G      2.039      1.198       1.86         23        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377      0.801      0.562      0.627      0.402

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     21/100      17.4G      2.004      1.215      1.849          4        640: 100%|██████████| 36/36 [00:16<00:00,  2.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.793      0.568      0.635      0.402

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     22/100      17.4G      2.059      1.209      1.842          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.834      0.579      0.652      0.415

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     23/100      17.5G      1.978      1.146      1.829         27        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.80it/s]


                   all       1274       7377      0.835      0.578      0.655      0.412

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     24/100      17.4G      1.963      1.112       1.82          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.86it/s]


                   all       1274       7377      0.819       0.59      0.659      0.422

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     25/100      17.4G      1.957       1.13      1.823         20        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.83it/s]


                   all       1274       7377      0.831       0.57      0.646      0.417

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     26/100      17.5G      1.928        1.1       1.84         18        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.818        0.6      0.667      0.424

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     27/100      17.4G      2.003      1.123      1.824         11        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.78it/s]


                   all       1274       7377      0.828      0.588      0.663      0.423

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     28/100      17.3G      1.928      1.099      1.816         20        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.856      0.594      0.673      0.432

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     29/100      17.4G      1.883      1.047      1.798          7        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.843      0.589       0.67      0.436

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     30/100      17.3G      1.934      1.065      1.825         11        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.849      0.598       0.68      0.435

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     31/100      17.4G      1.884      1.052      1.803         19        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.94it/s]


                   all       1274       7377      0.833      0.592       0.67       0.43

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     32/100      17.4G      1.886      1.064      1.796         10        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.826      0.594      0.677      0.434

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     33/100      17.4G      1.852      1.015      1.789         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.824      0.611      0.682       0.44

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     34/100      17.4G      1.859      1.003       1.81         11        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.96it/s]


                   all       1274       7377      0.831      0.604      0.676      0.434

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     35/100      17.4G      1.904      1.036      1.817         21        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.72it/s]


                   all       1274       7377      0.845      0.608      0.695      0.454

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     36/100      17.4G      1.848      1.003      1.795          9        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.98it/s]


                   all       1274       7377       0.84      0.614      0.697      0.455

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     37/100      17.3G      1.851      1.002      1.815         10        640: 100%|██████████| 36/36 [00:16<00:00,  2.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.68it/s]


                   all       1274       7377      0.861      0.612      0.696      0.452

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     38/100      17.3G      1.781       1.08      1.755          0        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.96it/s]


                   all       1274       7377      0.857        0.6       0.69      0.451

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     39/100      17.4G      1.821      0.972      1.786         21        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.859      0.592      0.681      0.449

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     40/100      17.3G      1.813     0.9954      1.783         15        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.829      0.622      0.694      0.448

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     41/100      17.4G      1.775     0.9808      1.793         10        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.844      0.616      0.698      0.449

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     42/100      17.4G      1.812     0.9939      1.778         24        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.84it/s]


                   all       1274       7377      0.842      0.616      0.702      0.459

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     43/100      17.4G      1.769     0.9863        1.8          4        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.866      0.602      0.703      0.467

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     44/100      17.3G      1.812     0.9599      1.761         26        640: 100%|██████████| 36/36 [00:16<00:00,  2.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.846      0.624      0.711      0.468

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     45/100      17.4G      1.721     0.9111      1.774          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377       0.84      0.616      0.704      0.465

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     46/100      17.4G      1.713     0.9063      1.767          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377      0.885       0.61      0.714      0.471

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     47/100      17.3G      1.737     0.9066      1.759         15        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.865      0.611      0.702      0.463

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     48/100      17.3G      1.782     0.9158       1.76         16        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377       0.86      0.624      0.712      0.465

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     49/100      17.3G      1.773      1.076      1.737          7        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.831      0.638      0.707      0.465

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     50/100      17.3G      1.801      0.926      1.765         17        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.849      0.628      0.708      0.468

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     51/100      17.4G      1.741     0.9015      1.779         10        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.846      0.621      0.703      0.462

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     52/100      17.4G      1.703     0.8868      1.773          5        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.842      0.631      0.713      0.472

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     53/100      17.3G      1.695      0.884      1.774          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.92it/s]


                   all       1274       7377      0.852      0.625      0.715      0.477

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     54/100      17.3G      1.652     0.8591      1.751         14        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.854      0.636      0.718      0.475

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     55/100      17.4G      1.719     0.8931      1.761         21        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377       0.85      0.629      0.717      0.474

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     56/100      17.4G       1.64     0.8626      1.763          5        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.852      0.632      0.716      0.478

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     57/100      17.4G      1.697     0.8727      1.743         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.851      0.636      0.719      0.479

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     58/100      17.4G      1.675     0.8768      1.755         21        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.867      0.628      0.727      0.481

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     59/100      17.3G      1.685     0.8588      1.749         12        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.864      0.633      0.724      0.482

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     60/100      17.3G      1.671     0.8424      1.744         18        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377       0.87      0.633      0.728      0.481

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     61/100      17.3G      1.685      0.848      1.741         39        640: 100%|██████████| 36/36 [00:16<00:00,  2.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.77it/s]


                   all       1274       7377      0.859      0.638      0.728      0.484

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     62/100      17.3G      1.624      0.826       1.75         33        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.868      0.634      0.726      0.484

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     63/100      17.3G      1.678     0.8427      1.747         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.856      0.633      0.729      0.487

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     64/100      17.4G      1.683      0.838      1.727          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.24it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.856       0.64      0.725      0.487

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     65/100      17.4G      1.698     0.8532      1.746          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377       0.87      0.639      0.733      0.493

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     66/100      17.3G      1.679     0.8467      1.756         10        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.855       0.65      0.731      0.491

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     67/100      17.4G      1.686     0.8483      1.751         19        640: 100%|██████████| 36/36 [00:16<00:00,  2.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.69it/s]


                   all       1274       7377      0.851      0.646      0.731       0.49

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     68/100      17.4G      1.633     0.8128      1.716         19        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.98it/s]


                   all       1274       7377      0.863      0.644      0.732      0.489

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     69/100      17.5G      1.637     0.8179      1.736         16        640: 100%|██████████| 36/36 [00:16<00:00,  2.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377      0.861      0.646      0.737      0.496

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     70/100      17.4G      1.633     0.8081      1.737         24        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377      0.863      0.646      0.731      0.494

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     71/100      17.4G      1.604     0.7836      1.735         17        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.857      0.644      0.732      0.491

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     72/100      17.4G       1.61     0.7937      1.732         12        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.855      0.651      0.739      0.499

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     73/100      17.3G      1.627     0.8133      1.743         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.69it/s]


                   all       1274       7377      0.874      0.645      0.733      0.493

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     74/100      17.5G      1.573     0.7842      1.746         11        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.71it/s]


                   all       1274       7377      0.865      0.648      0.742        0.5

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     75/100      17.4G      1.599     0.7872      1.731          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.70it/s]


                   all       1274       7377      0.861      0.651      0.741      0.502

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     76/100      17.4G      1.607     0.7845      1.722          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377       0.87      0.652      0.739      0.499

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     77/100      17.4G      1.615     0.7972      1.725         22        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.873      0.649      0.744      0.502

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     78/100      17.4G      1.567     0.7508      1.733          4        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.886      0.643      0.744      0.506

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     79/100      17.3G      1.566     0.7642      1.728         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.859      0.654      0.746      0.504

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     80/100      17.4G      1.549      0.771      1.718         36        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377       0.88      0.648      0.746      0.503

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     81/100      17.4G      1.548     0.7844      1.729          7        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.852      0.663      0.745      0.503

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     82/100      17.5G       1.55     0.7648      1.724          9        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.02it/s]


                   all       1274       7377      0.867      0.659      0.744      0.504

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     83/100      17.4G       1.55     0.7577      1.725          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.854       0.66      0.748      0.508

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     84/100      17.3G      1.554     0.7669      1.721          8        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.848      0.661      0.747      0.505

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     85/100      17.6G      1.558     0.7593      1.729         11        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.83it/s]


                   all       1274       7377      0.867      0.659      0.748      0.509

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     86/100      17.4G      1.622      1.054      1.848          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.867      0.652      0.746      0.509

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     87/100      17.3G      1.547     0.7509      1.722         14        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.91it/s]


                   all       1274       7377      0.864      0.661      0.752      0.512

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     88/100      17.4G      1.546     0.7337      1.715         16        640: 100%|██████████| 36/36 [00:16<00:00,  2.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.03it/s]


                   all       1274       7377      0.873      0.655      0.749      0.511

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     89/100      17.3G      1.451     0.6907      1.722         14        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.874      0.654      0.747       0.51

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     90/100      17.4G      1.484      0.744      1.741          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.885      0.645      0.749       0.51
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     91/100      17.4G      1.612     0.7201      1.688         13        640: 100%|██████████| 36/36 [00:17<00:00,  2.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.861      0.658      0.746       0.51

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     92/100      17.4G      1.626     0.7385      1.699         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.95it/s]


                   all       1274       7377      0.874      0.656      0.745      0.507

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     93/100      17.4G      1.612     0.7087      1.688          7        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.861      0.661       0.75      0.513

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     94/100      17.4G      1.614     0.7366       1.69          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.866      0.663      0.753      0.515

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     95/100      17.5G      1.639     0.7385       1.68         31        640: 100%|██████████| 36/36 [00:16<00:00,  2.21it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.72it/s]


                   all       1274       7377      0.862      0.663      0.749      0.512

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     96/100      17.4G      1.605     0.7219      1.688         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.12it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.862      0.661       0.75      0.514

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     97/100      17.4G      1.597     0.7149      1.686         13        640: 100%|██████████| 36/36 [00:16<00:00,  2.17it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.862      0.663      0.754      0.516

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     98/100      17.3G      1.591     0.7021      1.664          3        640: 100%|██████████| 36/36 [00:16<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.01it/s]


                   all       1274       7377      0.862      0.662      0.755      0.517

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     99/100      17.3G      1.575     0.6978      1.676         40        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  3.00it/s]


                   all       1274       7377      0.856      0.666      0.757      0.518

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


    100/100      17.4G      1.566      0.685      1.677          6        640: 100%|██████████| 36/36 [00:16<00:00,  2.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.99it/s]


                   all       1274       7377      0.877       0.66      0.756      0.516

100 epochs completed in 0.872 hours.
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/last.pt, 33.5MB
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt, 33.5MB

Validating /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt...
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:1 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:2 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:3 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:09<00:00,  2.20it/s]


                   all       1274       7377      0.858      0.665      0.757      0.518
            Pedestrian        333        600       0.78       0.69      0.757      0.468
         Traffic_Signs        309        434      0.868       0.62      0.723       0.45
               Vehicle       1034       2820      0.923      0.889      0.942      0.772
         traffic_light        660       3523      0.861      0.462      0.605      0.383
Speed: 0.1ms preprocess, 2.9ms inference, 0.0ms loss, 0.3ms postprocess per image
Results saved to /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train


# Results on Validation set:
100 epochs completed in 0.872 hours.
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/last.pt, 33.5MB
Optimizer stripped from /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt, 33.5MB

Validating /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt...
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:1 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:2 (NVIDIA RTX A6000, 48677MiB)
                                                           CUDA:3 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:09<00:00,  2.20it/s]
                   all       1274       7377      0.858      0.665      0.757      0.518
            Pedestrian        333        600       0.78       0.69      0.757      0.468
         Traffic_Signs        309        434      0.868       0.62      0.723       0.45
               Vehicle       1034       2820      0.923      0.889      0.942      0.772
         traffic_light        660       3523      0.861      0.462      0.605      0.383
Speed: 0.1ms preprocess, 2.9ms inference, 0.0ms loss, 0.3ms postprocess per image
Results saved to /home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train

# Test the results on Test dataset
- The above results was on the validation set of the Carla_labeling-v5 dataset (See the Image counts)
- Now we need to calculate the metrics on the test set of the Carla_labeling-v5 dataset containing only 641 images.
- Below is the script to do so:

In [6]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/last.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.25, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:10<00:00,  1.72s/it]


                   all        641       3701      0.852      0.644      0.779      0.572
            Pedestrian        182        332      0.775      0.642      0.757      0.507
         Traffic_Signs        168        230      0.839      0.613      0.762      0.513
               Vehicle        528       1377       0.94      0.883      0.935       0.81
         traffic_light        338       1762      0.855      0.439      0.664      0.458
Speed: 0.3ms preprocess, 6.1ms inference, 0.0ms loss, 0.1ms postprocess per image
Saving runs/detect/val/predictions.json...
Results saved to runs/detect/val


# Results on Last checpoint (100th epoch):
Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs
val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:10<00:00,  1.72s/it]
                   all        641       3701      0.852      0.644      0.779      0.572
            Pedestrian        182        332      0.775      0.642      0.757      0.507
         Traffic_Signs        168        230      0.839      0.613      0.762      0.513
               Vehicle        528       1377       0.94      0.883      0.935       0.81
         traffic_light        338       1762      0.855      0.439      0.664      0.458
Speed: 0.3ms preprocess, 6.1ms inference, 0.0ms loss, 0.1ms postprocess per image
Saving runs/detect/val/predictions.json...
Results saved to runs/detect/val

# Results on Test Set using Best Checkpoint:

In [7]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load the model
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.25, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

Ultralytics YOLOv8.2.93 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:11<00:00,  1.85s/it]


                   all        641       3701      0.858      0.642       0.78      0.572
            Pedestrian        182        332      0.768      0.639      0.753      0.503
         Traffic_Signs        168        230       0.86      0.613      0.766      0.514
               Vehicle        528       1377      0.937      0.882      0.934       0.81
         traffic_light        338       1762      0.867      0.434      0.665       0.46
Speed: 0.4ms preprocess, 5.9ms inference, 0.0ms loss, 0.1ms postprocess per image
Saving runs/detect/val2/predictions.json...
Results saved to runs/detect/val2


In [15]:
print(checkpoint.keys())

dict_keys(['date', 'version', 'license', 'docs', 'epoch', 'best_fitness', 'model', 'ema', 'updates', 'optimizer', 'train_args', 'train_metrics', 'train_results'])


# Taking the Optimal value from the F1-Confidence curve for Optimal Confidence
- Optimal Confidence YOLOv10m - 0.32
- IoU threshold : 0.5
- Results save d at: runs/detect/val3

In [1]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/last.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.32, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.5, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:07<00:00,  1.33s/it]


                   all        641       3701      0.876      0.623      0.775      0.576
            Pedestrian        182        332      0.806      0.602      0.748      0.512
         Traffic_Signs        168        230      0.866      0.591      0.758      0.519
               Vehicle        528       1377      0.952      0.877      0.933      0.811
         traffic_light        338       1762       0.88       0.42      0.663      0.461
Speed: 1.7ms preprocess, 5.1ms inference, 0.0ms loss, 0.1ms postprocess per image
Saving runs/detect/val3/predictions.json...
Results saved to runs/detect/val3


# Best Checkpoint + Taking the Optimal value from the F1-Confidence curve for Optimal Confidence
- Optimal Confidence YOLOv10m - 0.32
- IoU threshold : 0.5
- Results save d at: runs/detect/val4

In [1]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.32, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.5, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:07<00:00,  1.29s/it]


                   all        641       3701      0.879      0.618      0.773      0.575
            Pedestrian        182        332       0.81      0.602      0.747      0.509
         Traffic_Signs        168        230      0.868      0.574      0.751      0.517
               Vehicle        528       1377      0.956      0.875      0.932       0.81
         traffic_light        338       1762      0.882      0.419      0.662      0.462
Speed: 1.7ms preprocess, 5.3ms inference, 0.0ms loss, 0.1ms postprocess per image
Saving runs/detect/val4/predictions.json...
Results saved to runs/detect/val4


# Best Checkpoint + Optimal Confidence + IoU = 0.6
- Optimal Confidence YOLOv10m - 0.32
- IoU threshold : 0.6
- Results save d at: runs/detect/val4
- Best epoch = 99

In [1]:
from ultralytics import YOLO

# Load YOLOv10s model from scratch
# model = YOLO("yolov10s.pt")
# Load YOLOv10s Best weights on carla dataset
model = YOLO("/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt")

# Train the model: https://docs.ultralytics.com/modes/train/#train-settings
validation_results = model.val(data="/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml", 
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.32, # Sets the minimum confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="test",              
)

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/test/labels.cache... 641 images, 53 backgrounds, 0 corrupt: 100%|██████████| 641/641 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:08<00:00,  1.34s/it]


                   all        641       3701      0.879      0.618      0.773      0.575
            Pedestrian        182        332       0.81      0.602      0.747      0.509
         Traffic_Signs        168        230      0.868      0.574      0.751      0.517
               Vehicle        528       1377      0.956      0.875      0.932       0.81
         traffic_light        338       1762      0.882      0.419      0.662      0.462
Speed: 1.9ms preprocess, 5.3ms inference, 0.0ms loss, 0.3ms postprocess per image
Saving runs/detect/val5/predictions.json...
Results saved to runs/detect/val5


# Seperate evaluation on the training set

In [1]:
from ultralytics import YOLO

# Load your model
model = YOLO('/home/sur06423/project/Version_5_exp/YOLO_v10/v10_m/train/weights/best.pt')  # replace with your model path

# Validate on custom dataset
results = model.val(data='/home/sur06423/project/Version_5_exp/Carla_Labeling-5/data.yaml',
            batch=128, 
            imgsz=640,
            save_json = True, # If True, saves the results to a JSON file for further analysis or integration with other tools.
            conf=0.25, # Sets the optimal confidence threshold for detections. Detections with confidence below this threshold are discarded.
            iou=0.6, # Sets the Intersection Over Union (IoU) threshold for Non-Maximum Suppression (NMS). Helps in reducing duplicate detections.
            device=0, 
            split="train", 
)  # Ensure your YAML points to your desired dataset

WARNING ⚠️ Python>=3.10 is required, but Python==3.9.19 is currently installed 
Ultralytics 8.3.13 🚀 Python-3.9.19 torch-2.4.1+cu121 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
YOLOv10m summary (fused): 369 layers, 16,455,016 parameters, 0 gradients, 63.4 GFLOPs


val: Scanning /home/sur06423/project/Version_5_exp/Carla_Labeling-5/train/labels.cache... 4485 images, 430 backgrounds, 0 corrupt: 100%|██████████| 4485/4485 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 36/36 [00:35<00:00,  1.02it/s]


                   all       4485      26788      0.893      0.684      0.816      0.638
            Pedestrian       1229       2227      0.848       0.74      0.838      0.624
         Traffic_Signs       1093       1459      0.891      0.649      0.798      0.581
               Vehicle       3636      10031      0.948      0.912      0.951      0.845
         traffic_light       2373      13071      0.886      0.433      0.677      0.503
Speed: 0.4ms preprocess, 4.8ms inference, 0.0ms loss, 0.1ms postprocess per image
Saving runs/detect/val6/predictions.json...
Results saved to runs/detect/val6


# This notebook is complete.
- All the cell are executed and saved for later use.
- Results are saved in the corresponding directories.